[JSC] Expand IC handler coverage for InHit / InMiss / InstanceOfHit / InstanceOfMiss

Constellation opened the pull request:
#### adef555a17500963f196a68743515343a9a574ad
<pre>
[JSC] Expand IC handler coverage for InHit / InMiss / InstanceOfHit / InstanceOfMiss
<a href="https://bugs.webkit.org/show_bug.cgi?id=275509">https://bugs.webkit.org/show_bug.cgi?id=275509</a>
<a href="https://rdar.apple.com/129867837">rdar://129867837</a>

Reviewed by Yijia Huang.

This patch expands IC handler coverage. Adding InHit / InMiss / InstanceOfHit / InstanceOfMiss support.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheHandler::createPreCompiled):
(JSC::inByIdInHandlerImpl):
(JSC::inByIdHitHandler):
(JSC::inByIdMissHandler):
(JSC::instanceOfHandlerImpl):
(JSC::instanceOfHitHandler):
(JSC::instanceOfMissHandler):
(JSC::InlineCacheCompiler::compileOneAccessCaseHandler):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.h:
* Source/JavaScriptCore/jit/BaselineJITRegisters.h:
* Source/JavaScriptCore/jit/JITThunks.h:

Canonical link: <a href="https://commits.webkit.org/280040@main">https://commits.webkit.org/280040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f719a99c4f00225ae0e280799f8eaffab794de5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55562 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34885 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58546 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5992 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42507 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6191 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/44744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4114 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57591 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32789 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47888 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25871 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29573 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5217 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4136 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48639 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51506 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5484 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60137 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54799 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30716 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5609 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/52180 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/31801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47958 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32882 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/67061 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8192 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/31548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12770 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -5909,6 +6001,51 @@ AccessGenerationResult InlineCacheCompiler::compileOneAccessCaseHandler(CodeBloc
+            case AccessType::InById: {
+                switch (accessCase.m_type) {
+                case AccessCase::InHit:
+                case AccessCase::InMiss: {
```
Here is my understanding of `InHit` and `InMiss` cases in ICs in JSC.  They are the optimizations for `in` operator in JavaScript. For example:
```
let obj = { foo: 42 };
console.log('foo' in obj); // true
console.log('bar' in obj); // false
```

To optimize this, this patch uses ICs to remember the results of previous in operations and use that information to speed up future checks.

`InHit`: 
1. This is a case where the `in` check succeeds, meaning the property is found in the object or its prototype chain.
2. When the `in` check is performed and finds the property, the result is cached as an `InHit`.
3. Subsequent checks for the same property can quickly return a "hit" without redoing the full property search.

`InMiss`: 
1. This is a case where the `in` check fails, meaning the property is not found in the object or its prototype chain.
2. When the `in` check does not find the property, the result is cached as an InMiss.
3. Subsequent checks for the same missing property can quickly return a "miss" avoiding the full search again.

hyjorc1 replied:
The actual structure checks are managed by the JIT code in `InByIdHitHandler ` and `InByIdMissHandler `.

hyjorc1 commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -1558,6 +1558,11 @@ Ref<InlineCacheHandler> InlineCacheHandler::createPreCompiled(Ref<InlineCacheHan
     }
+    case AccessCase::InstanceOfHit:
+    case AccessCase::InstanceOfMiss: {
+        result->u.s1.m_holder = accessCase.as<InstanceOfAccessCase>().prototype();
```
The handling for `InHit`/`InMiss` is simpler because it involves checking if the property exists. So, no need to store additional info here.

hyjorc1 replied:
For `InstanceOfHit` and `InstanceOfMiss`, we set to the prototype that the instanceof check is validating against. This allows the handler to quickly verify the prototype during the instanceof operation.

hyjorc1 commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -5909,6 +6001,51 @@ AccessGenerationResult InlineCacheCompiler::compileOneAccessCaseHandler(CodeBloc
+            case AccessType::InstanceOf: {
+                switch (accessCase.m_type) {
+                case AccessCase::InstanceOfHit:
+                case AccessCase::InstanceOfMiss: {
```
So, `A InstanceOf B` is basically checking whether `A.prototype` existing in the prototype chain of `B`. And this can also be optimized by ICs.

`InstanceOfHit`: This case is used when the `instanceof` check is true, meaning the prototype of the constructor is found in the object's prototype chain.

`InstanceOfMiss`: This case is used when the `instanceof` check is false, meaning the prototype of the constructor is not found in the object's prototype chain.

When an `instanceof` check is performed and the result is determined (either hit or miss), the result is cached. This allows subsequent instanceof checks for the same objects and constructor to quickly return the cached result.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Expand IC handler coverage for InHit / InMiss / InstanceOfHit / InstanceOfMiss
https://bugs.webkit.org/show_bug.cgi?id=275509
rdar://129867837

Reviewed by Yijia Huang.

This patch expands IC handler coverage. Adding InHit / InMiss / InstanceOfHit / InstanceOfMiss support.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheHandler::createPreCompiled):
(JSC::inByIdInHandlerImpl):
(JSC::inByIdHitHandler):
(JSC::inByIdMissHandler):
(JSC::instanceOfHandlerImpl):
(JSC::instanceOfHitHandler):
(JSC::instanceOfMissHandler):
(JSC::InlineCacheCompiler::compileOneAccessCaseHandler):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.h:
* Source/JavaScriptCore/jit/BaselineJITRegisters.h:
* Source/JavaScriptCore/jit/JITThunks.h:

Canonical link: https://commits.webkit.org/280040@main
